src/math_tools/ipartial_derivatives.hpp

```diff
@@ -10,43 +10,83 @@
-    using IdxRangeFieldVal = typename DFieldValue::discrete_domain_type;
-
     /// The type of the object that will be differentiated.
-    using DFieldVal = DFieldValue;
+    using DFieldType = DField<IdxRangeType>;
+    using DFieldMemType = DFieldMem<IdxRangeType>;
 
     /// The type of the calculated derivative.
-    using DConstFieldVal = typename DFieldValue::view_type;
+    using DConstFieldType = typename DFieldType::view_type;
 
     /// The type of the grid on the dimension Xi on which the partial derivative is calculated.
-    using GridDerivativeDirection
-            = find_grid_t<DerivativeDirection, ddc::to_type_seq_t<IdxRangeFieldVal>>;
+    using GridDerivativeDim = find_grid_t<DerivativeDim, ddc::to_type_seq_t<IdxRangeType>>;
 
     /// The index range of the dimension Xi on which the partial derivative is calculated.
-    using IdxRangeDeriv = IdxRange<GridDerivativeDirection>;
+    using IdxRangeDeriv = IdxRange<GridDerivativeDim>;
 
     /// The index range of all dimensions except Xi.
-    using IdxRangeBatch = ddc::remove_dims_of_t<IdxRangeFieldVal, GridDerivativeDirection>;
+    using IdxRangeBatched = ddc::remove_dims_of_t<IdxRangeType, GridDerivativeDim>;
 
 public:
     /**
     * @brief Compute the partial derivative of @f$ F(X1,..,Xn)@f$ in Xi direction.
     *
-    * @param[out] dfieldval_dxi Partial derivatives in Xi direction.
     * @param[in] fieldval Values of the field @f$ F(X1,..,Xn)@f$.
     */
-    virtual void operator()(DFieldVal dfieldval_dxi, DConstFieldVal fieldval) const = 0;
+    virtual DFieldType operator()(DConstFieldType fieldval) const = 0;
+};
+
+/**
+ * @brief A class which provides access to a partial derivative function which can be preallocated where useful.
+ *
+ * An abstract class which implements a preallocate function returning a pointer to an IPartialDerivative.
+ * A pointer to an IPartialDerivative is used so that the returned object can be any sub-class of IPartialDerivative.
+ * The type (and thus the implementation of the operator) will be determined when the pointer is
+ * dereferenced.
+ *
+ * The preallocate function should be used to allocate an instance of the IPartialDerivative before
+ * using it repeatedly. Once the preallocated object goes out of scope it will be deallocated.
+ * This means that objects of this class take up little or no space in memory.
+ *
+ */
+template <class IdxRangeType, class DerivativeDim>
+class IPartialDerivativeCreator : public IPartialDerivative<IdxRangeType, DerivativeDim>
+{
+public:
+    /**
+     * @brief Allocate an instance of a pointer to an IPartialDerivative.
+     *
+     * Allocate and return an instance of a pointer to a sub-class
+     * of IPartialDerivative.
+     *
+     * @return A pointer to an IPartialDerivative.
+     *
+     * @see IPartialDerivative
+     */
+    virtual std::unique_ptr<IPartialDerivative<IdxRangeType, DerivativeDim>> preallocate()
+            const = 0;
```

Comment:
I thought we had discussed a method:
```suggestion
    virtual std::unique_ptr<IPartialDerivative<IdxRangeType, DerivativeDim>> preallocate(DConstField<IdxRangeType> fieldval)
            const = 0;
```
If the preallocation does not require a field, what is the advantage of this class?

Reply:
Should be solved now